### ISSUE TITLE
error in unloading C objects

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#52]
Look back into problematic unloading in test-models.R

Also see if we want to make the unloading more easily accessible to users.

**Comment by paciorek** [comment]
Here's a reproducible example; it fails both on Mac and Linux.

code <- nimbleCode({
a ~ dnorm(0,1)
})
m <- nimbleModel(code)
cm <- compileNimble(m)
dyn.unload(getNimbleProject(m)$cppProjects[[1]]$getSOName())

**Comment by paciorek** [comment]
and one more tidbit: unloading a compiled RC function works fine (but unloading a compiled true nimbleFunction does not).

here's the RC function example that does not cause a seg fault when R exits:

nf <- nimbleFunction(
    run = function(p = double(0)) {
        returnType(double(0))
        out <- 3
        return(out)
    }
    )
project <- nimbleProjectClass(NULL, name = 'foo')
cnf <- compileNimble(nf, project = project)
dyn.unload(project$cppProjects[[1]]$getSOName())

**Comment by perrydv** [comment]
I think I have this fixed, or least less of a problem.  In the past we have generated a very short finalizer function for each generated C++ class so that when we register an R external pointer to an object of the class we can provide a pointer to the finalizer function (R_RegisterCFinalizerEx).  I think the problem had to do with those finalizer functions getting taken away by dyn.unload but then the finalizer trying to call them (by function pointer) and finding nothing there and hence crashing.  I realized that in practice all of our generated classes inherit from base classes defined in CppCode, so we could rely on a single finalizer function in CppCode (i.e. loaded with the package) that calls a virtual base class destructor.  I made this change, so we no longer generate derived-class-specific finalizers (although the code to do so is still in the general cppDef system).  However this still did not fix the segfaults, and I am not sure why.  I am wondering if in R's termination sequence it might unload shared libraries before trying to clear all variables, and hence run into the same problem.  So next I turned off the "onExit" flag to R_RegisterCFinalizerEx.  Now during R's termination there is no attempt to call finalizers, so the problem does not occur, and the segfaults seem to not happen, at least in the examples I was using.  I believe the allocated C++ objects will be cleared anyway as part of an OS cleaning up memory when R terminates.
